framework/src/actions/SetupMeshCompleteAction.C

```diff
@@ -49,8 +51,24 @@ SetupMeshCompleteAction::act()
   if (!_mesh)
     mooseError("No mesh file was supplied and no generation block was provided");
 
-  completeSetup(_mesh.get());
+  /**
+   * If possible we'd like to refine the mesh here before the equation systems
+   * are setup to avoid doing expensive projections. If however we are doing a
+   * file based restart and we need uniform refinements, we'll have to postpone
+   * those refinements until after the solution has been read in.
+   */
+  if (_current_task == "uniform_refine_mesh" && _app.setFileRestart() == false)
```

Comment:
Cool: this will save a ton of time on bigger problems (especially on lots of processors)